File: mzcd3-code/code/iCloud/PPRecipes/PPRMasterViewController.m
/***
 * Excerpted from "Core Data in Objective-C, Third Edition",
 * published by The Pragmatic Bookshelf.
 * Copyrights apply to this code. It may not be used to create training material,
 * courses, books, articles, and the like. Contact us if you are in doubt.
 * We make no guarantees that this code is fit for any purpose.
 * Visit http://www.pragmaticprogrammer.com/titles/mzcd3 for more book information.
***/
//
//  PPRMasterViewController.m
//  PPRecipes
//
//  Created by Marcus Zarra on 4/3/12.
//  Copyright (c) 2012 The Pragmatic Programmer. All rights reserved.
//

#import "PPRMasterViewController.h"

#import "PPRDetailViewController.h"
#import "PPRDataController.h"

@interface PPRMasterViewController ()

- (void)configureCell:(UITableViewCell *)cell atIndexPath:(NSIndexPath *)indexPath;

@end

@implementation PPRMasterViewController

- (void)awakeFromNib
{
  if ([[UIDevice currentDevice] userInterfaceIdiom] == UIUserInterfaceIdiomPad) {
    [self setClearsSelectionOnViewWillAppear:NO];
    [self setContentSizeForViewInPopover:CGSizeMake(320.0, 600.0)];
  }
  [super awakeFromNib];
}

- (void)viewDidLoad
{
  [super viewDidLoad];

  [[self navigationItem] setLeftBarButtonItem:[self editButtonItem]];
  
  UIBarButtonItem *addButton = [[UIBarButtonItem alloc] initWithBarButtonSystemItem:UIBarButtonSystemItemAdd target:self action:@selector(insertNewObject:)];
  [[self navigationItem] setRightBarButtonItem:addButton];
  id detailVC = [[[[self splitViewController] viewControllers] lastObject] topViewController];
  [self setDetailViewController:detailVC];
}

- (void)viewDidUnload
{
  [super viewDidUnload];
}

- (BOOL)shouldAutorotateToInterfaceOrientation:(UIInterfaceOrientation)interfaceOrientation
{
  if ([[UIDevice currentDevice] userInterfaceIdiom] == UIUserInterfaceIdiomPhone) {
    return (interfaceOrientation != UIInterfaceOrientationPortraitUpsideDown);
  } else {
    return YES;
  }
}

- (void)insertNewObject:(id)sender
{
  NSManagedObjectContext *context = [[self fetchedResultsController] managedObjectContext];
  NSEntityDescription *entity = [[[self fetchedResultsController] fetchRequest] entity];
  [NSEntityDescription insertNewObjectForEntityForName:[entity name] inManagedObjectContext:context];
}

#pragma mark - Table View

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView
{
  return [[[self fetchedResultsController] sections] count];
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section
{
  id <NSFetchedResultsSectionInfo> sectionInfo = [[[self fetchedResultsController] sections] objectAtIndex:section];
  return [sectionInfo numberOfObjects];
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath
{
  UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:@"Cell"];
  [self configureCell:cell atIndexPath:indexPath];
  return cell;
}

- (BOOL)tableView:(UITableView *)tableView canEditRowAtIndexPath:(NSIndexPath *)indexPath
{
  return YES;
}

- (void)tableView:(UITableView *)tableView commitEditingStyle:(UITableViewCellEditingStyle)editingStyle forRowAtIndexPath:(NSIndexPath *)indexPath
{
  if (editingStyle != UITableViewCellEditingStyleDelete) return;
  
  NSManagedObjectContext *context = [self.fetchedResultsController managedObjectContext];
  [context deleteObject:[self.fetchedResultsController objectAtIndexPath:indexPath]];
}

- (BOOL)tableView:(UITableView *)tableView canMoveRowAtIndexPath:(NSIndexPath *)indexPath
{
  return NO;
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath
{
  if ([[UIDevice currentDevice] userInterfaceIdiom] != UIUserInterfaceIdiomPad) return;
  
  NSManagedObject *object = [[self fetchedResultsController] objectAtIndexPath:indexPath];
  [[self detailViewController] setDetailItem:object];
}

- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender
{
  if (![[segue identifier] isEqualToString:@"showDetail"]) return;
  
  NSIndexPath *indexPath = [self.tableView indexPathForSelectedRow];
  NSManagedObject *object = [[self fetchedResultsController] objectAtIndexPath:indexPath];
  [[segue destinationViewController] setDetailItem:object];
}

#pragma mark - Fetched results controller

- (NSFetchedResultsController *)fetchedResultsController
{
  if (_fetchedResultsController) return _fetchedResultsController;
  
  NSFetchRequest *fetchRequest = [[NSFetchRequest alloc] initWithEntityName:@"Recipe"];
  [fetchRequest setFetchBatchSize:20];
  
  [fetchRequest setSortDescriptors:[NSArray arrayWithObject:[[NSSortDescriptor alloc] initWithKey:@"name" ascending:YES]]];

  NSManagedObjectContext *moc = [[self dataController] managedObjectContext];
  _fetchedResultsController = [[NSFetchedResultsController alloc] initWithFetchRequest:fetchRequest managedObjectContext:moc sectionNameKeyPath:nil cacheName:@"Master"];
  [_fetchedResultsController setDelegate:self];

	NSError *error = nil;
	ZAssert([_fetchedResultsController performFetch:&error], @"Unresolved error %@\n%@", [error localizedDescription], [error userInfo]);
  
  return _fetchedResultsController;
}    

- (void)controllerWillChangeContent:(NSFetchedResultsController *)controller
{
  [[self tableView] beginUpdates];
}

- (void)controller:(NSFetchedResultsController *)controller didChangeSection:(id <NSFetchedResultsSectionInfo>)sectionInfo
           atIndex:(NSUInteger)sectionIndex forChangeType:(NSFetchedResultsChangeType)type
{
  switch(type) {
    case NSFetchedResultsChangeInsert:
      [[self tableView] insertSections:[NSIndexSet indexSetWithIndex:sectionIndex] withRowAnimation:UITableViewRowAnimationFade];
      break;
      
    case NSFetchedResultsChangeDelete:
      [[self tableView] deleteSections:[NSIndexSet indexSetWithIndex:sectionIndex] withRowAnimation:UITableViewRowAnimationFade];
      break;
    case NSFetchedResultsChangeMove:
    case NSFetchedResultsChangeUpdate:
      break;
  }
}

- (void)controller:(NSFetchedResultsController *)controller didChangeObject:(id)anObject
       atIndexPath:(NSIndexPath *)indexPath forChangeType:(NSFetchedResultsChangeType)type
      newIndexPath:(NSIndexPath *)newIndexPath
{
  switch(type) {
    case NSFetchedResultsChangeInsert:
      [[self tableView] insertRowsAtIndexPaths:[NSArray arrayWithObject:newIndexPath] withRowAnimation:UITableViewRowAnimationFade];
      break;
      
    case NSFetchedResultsChangeDelete:
      [[self tableView] deleteRowsAtIndexPaths:[NSArray arrayWithObject:indexPath] withRowAnimation:UITableViewRowAnimationFade];
      break;
      
    case NSFetchedResultsChangeUpdate:
      [self configureCell:[[self tableView] cellForRowAtIndexPath:indexPath] atIndexPath:indexPath];
      break;
      
    case NSFetchedResultsChangeMove:
      [[self tableView] deleteRowsAtIndexPaths:[NSArray arrayWithObject:indexPath] withRowAnimation:UITableViewRowAnimationFade];
      [[self tableView] insertRowsAtIndexPaths:[NSArray arrayWithObject:newIndexPath]withRowAnimation:UITableViewRowAnimationFade];
      break;
  }
}

- (void)controllerDidChangeContent:(NSFetchedResultsController *)controller
{
  [[self tableView] endUpdates];
}

- (void)configureCell:(UITableViewCell *)cell atIndexPath:(NSIndexPath *)indexPath
{
  NSManagedObject *object = [self.fetchedResultsController objectAtIndexPath:indexPath];
  [[cell textLabel] setText:[object valueForKey:@"name"]];
}

@end

File: mzswift-code/code/PPRecipes/PPRecipes/PPRMasterViewController.swift
//
//  PPREntryViewController.swift
//  PPRecipes
//
//  Created by Marcus S. Zarra on 12/6/15.
//  Copyright © 2015 The Pragmatic Programmer. All rights reserved.
//

import Foundation
import UIKit
import CoreData

class PPRMasterViewController: UITableViewController, NSFetchedResultsControllerDelegate {
  var managedObjectContext: NSManagedObjectContext?
  var fResultsController: NSFetchedResultsController<NSManagedObject>?
  var detailController: PPRDetailViewController?

  override func viewDidLoad() {
    super.viewDidLoad()

    navigationItem.leftBarButtonItem = editButtonItem

    if let navController = splitViewController?.viewControllers.last as? UINavigationController {
      detailController = navController.topViewController as? PPRDetailViewController
    }

    let fetch = NSFetchRequest<NSManagedObject>(entityName: "Recipe")
    fetch.sortDescriptors = [NSSortDescriptor(key: "name", ascending: true)]

    guard let moc = managedObjectContext else {
      fatalError("MOC not initialized")
    }
    fResultsController = NSFetchedResultsController(fetchRequest: fetch,
      managedObjectContext: moc, sectionNameKeyPath: nil, cacheName: nil)
    fResultsController?.delegate = self
    do {
      try fResultsController?.performFetch()
    } catch {
      fatalError("Unable to fetch: \(error)")
    }
  }

  func frcExmaple2() {
    let fetch = NSFetchRequest<NSManagedObject>(entityName: "Recipe")
    fetch.sortDescriptors = [NSSortDescriptor(key: "type", ascending: true),
      NSSortDescriptor(key: "name", ascending: true)]

    guard let moc = managedObjectContext else {
      fatalError("MOC not initialized")
    }
    fResultsController = NSFetchedResultsController(fetchRequest: fetch,
      managedObjectContext: moc, sectionNameKeyPath: "type",
      cacheName: "Master")
    fResultsController?.delegate = self
    do {
      try fResultsController?.performFetch()
    } catch {
      fatalError("Unable to fetch: \(error)")
    }
  }

  func prepareForDetailSegue(_ segue: UIStoryboardSegue) {
    let control = segue.destination as! PPRDetailViewController
    let path = tableView.indexPathForSelectedRow!
    let recipe = fResultsController!.object(at: path) as! PPRRecipeMO
    control.recipeMO = recipe
  }

  func prepareForAddRecipeSegue(_ segue: UIStoryboardSegue) {
    let tempController = segue.destination
    let controller = tempController as! PPREditRecipeViewController
    let moc = managedObjectContext!
    let recipe = NSEntityDescription.insertNewObject(forEntityName: "Recipe",
      into: moc) as! PPRRecipeMO
    controller.recipeMO = recipe
  }

  override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
    guard let identifier = segue.identifier else {
      fatalError("Unidentified segue")
    }

    switch identifier {
    case "showRecipe":
      prepareForDetailSegue(segue)
    case "addRecipe":
      prepareForAddRecipeSegue(segue)
    default:
      fatalError("Unexpected segue: \(identifier)")
    }
  }

  //MARK UITableViewDelegate

  override func numberOfSections(in tableView: UITableView) -> Int {
    guard let count = fResultsController?.sections?.count else {
      fatalError("Failed to resolve FRC")
    }
    return count
  }

  override func tableView(_ tableView: UITableView,
    numberOfRowsInSection section: Int) -> Int {
    guard let sectionInfo = fResultsController?.sections?[section] else {
      fatalError("Failed to resolve FRC")
    }
    return sectionInfo.numberOfObjects
  }

  override func tableView(_ tableView: UITableView,
    canEditRowAt indexPath: IndexPath) -> Bool {
    return true
  }

  override func tableView(_ tableView: UITableView,
    cellForRowAt indexPath: IndexPath) -> UITableViewCell {
    guard let frc = fResultsController else {
      fatalError("Failed to resolve NSFetchedResultsController")
    }
    let obj = frc.object(at: indexPath)
    guard let cell = tableView.dequeueReusableCell(withIdentifier: "Cell")
        else {
      fatalError("Failed to dequeue cell")
    }
    cell.textLabel?.text = obj.value(forKey: "name") as? String
    return cell
  }

  //MARK: NSFetchedResultsControllerDelegate

  func controllerWillChangeContent(_ controller:
    NSFetchedResultsController<NSFetchRequestResult>) {
    tableView.beginUpdates()
  }

  func controller(_ control: NSFetchedResultsController<NSFetchRequestResult>,
    didChange sectionInfo: NSFetchedResultsSectionInfo,
    atSectionIndex sectionIndex: Int, 
	for type: NSFetchedResultsChangeType) {
    switch type {
    case .insert:
      tableView.insertSections(IndexSet(integer: sectionIndex),
        with: .fade)
    case .delete:
      tableView.deleteSections(IndexSet(integer: sectionIndex),
        with: .fade)
    case .move: break
    case .update: break
    }
  }

  func controller(_ control: NSFetchedResultsController<NSFetchRequestResult>,
    didChange anObject: Any, at indexPath: IndexPath?,
    for type: NSFetchedResultsChangeType,
    newIndexPath: IndexPath?) {
    switch type {
    case .insert:
      guard let nip = newIndexPath else { fatalError("How?") }
      tableView.insertRows(at: [nip], with: .fade)
    case .delete:
      guard let ip = indexPath else { fatalError("How?") }
      tableView.deleteRows(at: [ip], with: .fade)
    case .move:
      guard let nip = newIndexPath else { fatalError("How?") }
      guard let ip = indexPath else { fatalError("How?") }
      tableView.deleteRows(at: [ip], with: .fade)
      tableView.insertRows(at: [nip], with: .fade)
    case .update:
      guard let ip = indexPath else { fatalError("How?") }
      tableView.reloadRows(at: [ip], with: .fade)
    }
  }

  func controller(_ control: NSFetchedResultsController<NSFetchRequestResult>,
    sectionIndexTitleForSectionName sectionName: String) -> String? {
    return "[\(sectionName)]"
  }

  func controllerDidChangeContent(_ controller:
    NSFetchedResultsController<NSFetchRequestResult>) {
    tableView.endUpdates()
  }
}
